File: algorithms/sorting/quicksort/javascript/quicksort.js
const quicksort = (array) => {
  if (array.length < 2) return array;

  const pivot = array[0];
  const smallerThanPivot = array.slice(1).filter(e => e <= pivot);
  const greaterThanPivot = array.slice(1).filter(e => e > pivot);

  return quicksort(smallerThanPivot).concat([pivot], quicksort(greaterThanPivot));
}

module.exports = { quicksort };

File: algorithms/sorting/quicksort/ruby/quicksort.rb
def quicksort(array)
  return array if array.size < 2

  pivot = array[0]
  smaller_than_pivot = array[1..-1].select { |e| e <= pivot}
  greater_than_pivot = array[1..-1].select { |e| e > pivot}

  quicksort(smaller_than_pivot) + [pivot] + quicksort(greater_than_pivot)
end

# Alternate approach choosing the last element as the pivot
def alternate_quicksort(array)
  return array if array.size < 2

  left = []
  right = []

  pivot_index = array.size - 1
  pivot_value = array[pivot_index]

  array.pop

  array.each do |item|
    item < pivot_value ? left.push(item) : right.push(item)
  end

  quicksort(left) + [pivot_value] + quicksort(right)
end
